File: scripts/debug_identity.jl
using Lux
using SafeTensors
using NNlib

function debug_identity()
    tensors = SafeTensors.load_safetensors("verification_data/identity_debug.safetensors")
    x = reshape(tensors["input"], reverse(size(tensors["input"])))
    y_ref = reshape(tensors["output"], reverse(size(tensors["output"])))
    w = reshape(tensors["weight"], reverse(size(tensors["weight"]))) # (3,3,3,1,1)
    
    # Print where the 1.0 is in weights
    cart_inds = findall(x -> x > 0.5, w)
    println("Weight 1.0 index in Lux (Reshape): ", cart_inds)
    
    # Expected: Center of 3x3x3 is (2,2,2)
    
    layer = Conv((3,3,3), 1=>1, stride=1, pad=1, use_bias=false)
    ps = (weight=w,)
    y, _ = layer(x, ps, (;))
    
    diff = abs.(y .- y_ref)
    println("Max Diff: ", maximum(diff))
    
    # Find active output pixel
    out_inds = findall(x -> x > 0.5, y)
    println("Output active pixel: ", out_inds)
    
    ref_inds = findall(x -> x > 0.5, y_ref)
    println("Ref active pixel: ", ref_inds)
end

debug_identity()

File: scripts/debug_identity.py
import torch
import torch.nn as nn
from safetensors.torch import save_file

def generate_identity_debug():
    # 1x1x1 Convolution with known weights
    conv = nn.Conv3d(1, 1, kernel_size=3, stride=1, padding=1, bias=False)
    
    # Set weight to all zeros except center
    nn.init.constant_(conv.weight, 0.0)
    conv.weight.data[0, 0, 1, 1, 1] = 1.0 # Center pixel identity
    
    # Input: All zeros with a single 1.0 at [0,0,16,16,16]
    x = torch.zeros(1, 1, 32, 32, 32)
    x[0, 0, 16, 16, 16] = 1.0
    
    with torch.no_grad():
        y = conv(x)
        
    save_file({
        "input": x,
        "output": y,
        "weight": conv.weight
    }, "verification_data/identity_debug.safetensors")

if __name__ == "__main__":
    generate_identity_debug()
